src/lib/pubkey/curve448/ed448/ed448_internal.cpp

```diff
@@ -0,0 +1,317 @@
+                                                        0x433b80e18b00938e,
+                                                        0x12ae1af72ab66511,
+                                                        0xea6de324a3d3a464,
+                                                        0x9e146570470f1767,
+                                                        0x221d15a622bf36da,
+                                                        0x4f1970c66bed0ded};
+   constexpr std::array<const uint64_t, WORDS_448> y = {0x9808795bf230fa14,
+                                                        0xfdbd132c4ed7c8ad,
+                                                        0x3ad3ff1ce67c39c4,
+                                                        0x87789c1e05a0c2d7,
+                                                        0x4bea73736ca39840,
+                                                        0x8876203756c9c762,
+                                                        0x693f46716eb6bc24};
+   return Ed448Point(Gf448Elem(x), Gf448Elem(y));
+}
+
+std::array<uint8_t, ED448_LEN> Ed448Point::encode() const {
+   std::array<uint8_t, ED448_LEN> res_buf = {0};
+
+   // RFC 8032 5.2.2
+   //    All values are coded as octet strings, and integers are coded using
+   //    little-endian convention. [...]
+   //    First, encode the y-coordinate as a little-endian string of 57 octets.
+   //    The final octet is always zero.
+   y().to_bytes(std::span(res_buf).first<56>());
+
+   //    To form the encoding of the point, copy the least significant bit of
+   //    the x-coordinate to the most significant bit of the final octet.
+   res_buf.back() = (static_cast<uint8_t>(x().is_odd()) << 7);
+
+   return res_buf;
+}
+
+Ed448Point Ed448Point::operator+(const Ed448Point& other) const {
+   // RFC 8032 5.2.4. - Point Addition (Add)
+   const Gf448Elem A = m_z * other.m_z;
+   const Gf448Elem B = square(A);
+   const Gf448Elem C = m_x * other.m_x;
+   const Gf448Elem D = m_y * other.m_y;
+   const Gf448Elem E = (-Gf448Elem(MINUS_D)) * C * D;
+   const Gf448Elem F = B - E;
+   const Gf448Elem G = B + E;
+   const Gf448Elem H = (m_x + m_y) * (other.m_x + other.m_y);
+   const Gf448Elem X3 = A * F * (H - C - D);
+   const Gf448Elem Y3 = A * G * (D - C);
+   const Gf448Elem Z3 = F * G;
+
+   return Ed448Point(X3, Y3, Z3);
+}
+
+Ed448Point Ed448Point::double_point() const {
+   // RFC 8032 5.2.4. - Point Addition (Double)
+   const Gf448Elem B = square(m_x + m_y);
+   const Gf448Elem C = square(m_x);
+   const Gf448Elem D = square(m_y);
+   const Gf448Elem E = C + D;
+   const Gf448Elem H = square(m_z);
+   const Gf448Elem J = E - (H + H);
+   const Gf448Elem X3 = (B - E) * J;
+   const Gf448Elem Y3 = E * (C - D);
+   const Gf448Elem Z3 = E * J;
+
+   return Ed448Point(X3, Y3, Z3);
+}
+
+Ed448Point Ed448Point::scalar_mul(const Scalar448& s) const {
+   Ed448Point res(0, 1);
+
+   // Square and multiply (double and add) in constant time.
```

Comment:
Lots of room for optimization here if performance becomes an issue.

Reply:
If you have any optimizations in mind, let me know 👍 Otherwise, I think the performance is currently acceptable.

Reply:
Just that double-and-always-add is (literally) the worst case scenario for a multiplication algorithm since for scalar length `l` it performs `l` doublings and `l` additions. Easy optimization would be a fixed window with a constant time table lookup, eg a 4 bit window you instead do `l` doublings and `l/4` additions.

Fine to use this if you consider current performance ok.

Reply:
This sounds promising! I will leave it with the simple variant for now since the PR is already complex enough. Maybe we can speed it up in a follow-up PR.